- **Based on the findings and insights from the data exploration, we will now proceed with data preprocessing.**

In [1]:
# Importing Required Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Defining the path to the datasets folder
data_dir = "datasets"
dataset_path = os.path.join(data_dir, "UCI_Credit_Card.csv")

# Loading the CSV file into a DataFrame
df = pd.read_csv(dataset_path)

# Displaying the first few rows to confirm loading
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1    20000.0    2          2         1   24      2      2     -1     -1   
1   2   120000.0    2          2         2   26     -1      2      0      0   
2   3    90000.0    2          2         2   34      0      0      0      0   
3   4    50000.0    2          2         1   37      0      0      0      0   
4   5    50000.0    1          2         1   57     -1      0     -1      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...        0.0        0.0        0.0       0.0     689.0       0.0   
1  ...     3272.0     3455.0     3261.0       0.0    1000.0    1000.0   
2  ...    14331.0    14948.0    15549.0    1518.0    1500.0    1000.0   
3  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
4  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default.payment.next.month  
0       0.0       0.0       0.0                           1  
1    1000.0       0.0    2000.0                           1  
2    1000.0    1000.0    5000.0                           0  
3    1100.0    1069.0    1000.0                           0  
4    9000.0     689.0     679.0                           0  

[5 rows x 25 columns]

# **Renaming Column Names**

In [3]:
# Renaming Columns names for clarity and consistency
column_renames = {
    'ID': 'ID',
    'LIMIT_BAL': 'Credit_Limit',
    'SEX': 'Gender',
    'EDUCATION': 'Education_Level',
    'MARRIAGE': 'Marital_Status',
    'AGE': 'Age',
    'PAY_0': 'Sept_Pay',
    'PAY_2': 'Aug_Pay',
    'PAY_3': 'July_Pay',
    'PAY_4': 'June_Pay',
    'PAY_5': 'May_Pay',
    'PAY_6': 'Apr_Pay',
    'BILL_AMT1': 'Sept_Bill_Amt',
    'BILL_AMT2': 'Aug_Bill_Amt',
    'BILL_AMT3': 'July_Bill_Amt',
    'BILL_AMT4': 'June_Bill_Amt',
    'BILL_AMT5': 'May_Bill_Amt',
    'BILL_AMT6': 'Apr_Bill_Amt',
    'PAY_AMT1': 'Sept_Pay_Amt',
    'PAY_AMT2': 'Aug_Pay_Amt',
    'PAY_AMT3': 'July_Pay_Amt',
    'PAY_AMT4': 'June_Pay_Amt',
    'PAY_AMT5': 'May_Pay_Amt',
    'PAY_AMT6': 'Apr_Pay_Amt',
    'default.payment.next.month': 'default_payment_next_month'
}

# Renaming the columns in the DataFrame
df = df.rename(columns=column_renames)

# Verifying the new column names
df.columns

Index(['ID', 'Credit_Limit', 'Gender', 'Education_Level', 'Marital_Status',
       'Age', 'Sept_Pay', 'Aug_Pay', 'July_Pay', 'June_Pay', 'May_Pay',
       'Apr_Pay', 'Sept_Bill_Amt', 'Aug_Bill_Amt', 'July_Bill_Amt',
       'June_Bill_Amt', 'May_Bill_Amt', 'Apr_Bill_Amt', 'Sept_Pay_Amt',
       'Aug_Pay_Amt', 'July_Pay_Amt', 'June_Pay_Amt', 'May_Pay_Amt',
       'Apr_Pay_Amt', 'default_payment_next_month'],
      dtype='object')

In [4]:
# Defineing numerical columns based on dataset information
numerical_columns = ['Credit_Limit', 'Age', 'Sept_Bill_Amt', 'Aug_Bill_Amt', 'July_Bill_Amt', 'June_Bill_Amt', 'May_Bill_Amt', 'Apr_Bill_Amt', 'Sept_Pay_Amt', 'Aug_Pay_Amt', 'July_Pay_Amt', 'June_Pay_Amt', 'May_Pay_Amt', 'Apr_Pay_Amt']

# Defining categorical columns based on dataset information
categorical_columns = ['Gender', 'Education_Level', 'Marital_Status', 'Sept_Pay', 'Aug_Pay', 'July_Pay', 'June_Pay', 'May_Pay', 'Apr_Pay']

# Defining the target column
target_column = 'default_payment_next_month'

# **1. Handling Missing Values**
- We will handle the following data cleaning tasks based on the observations and assumptions from the data exploration phase:

In [5]:
#Handling missing values in Marital Status
df['Marital_Status'].value_counts()

Marital_Status
2    15964
1    13659
3      323
0       54
Name: count, dtype: int64

- 54 missing values in Marital Status which is very small compared to the total number of records (30000). We will replace these missing values with the mode of the column, which is 2 (Single).

In [6]:
# Calculating mode for Marital_Status
marital_mode = df['Marital_Status'].mode()[0]
print(f"\nMode of Marital Status: {marital_mode}")

# Replacing 0s in Marital_Status with mode
df['Marital_Status'] = df['Marital_Status'].replace(0, marital_mode)

# Verifying correction
df['Marital_Status'].value_counts()


Mode of Marital Status: 2


Marital_Status
2    16018
1    13659
3      323
Name: count, dtype: int64

- Replaced missing values '0' in Marital_Status with the mode (2-single).

In [7]:
#Handling missing values and inconsistencies in Education Level
print(f"Count of Education Level 0: {len(df[df['Education_Level'] == 0])}")
print(f"Count of Education Level 4 (others): {len(df[df['Education_Level'] == 4])}")
print(f"Count of Education Level 5 (unknown): {len(df[df['Education_Level'] == 5])}")
print(f"Count of Education Level 6 (unknown): {len(df[df['Education_Level'] == 6])}")

Count of Education Level 0: 14
Count of Education Level 4 (others): 123
Count of Education Level 5 (unknown): 280
Count of Education Level 6 (unknown): 51


In [8]:
# Replacing 4, 5, and 6 in Education Level with 0 and considering them as others
df['Education_Level'] = df['Education_Level'].replace([4,5,6], 0)

# Verifying correction
df['Education_Level'].value_counts()

Education_Level
2    14030
1    10585
3     4917
0      468
Name: count, dtype: int64

- Replaced 4, 5, and 6 in Education_Level with 0 and consider them as others. This is because these values are not standard education levels and can be grouped together.

In [9]:
# Handling Duplicate Rows
df.duplicated().sum()

np.int64(0)

- There are no duplicate rows in the dataset as we have `ID` column which is unique for each row.
- We will drop the `ID` column as it is not needed for modeling and will not contribute to the predictive power of the model.
- After dropping the ID column, we will once again check for duplicates to ensure data integrity.

In [10]:
# Dropping the ID column as it is not needed for modeling
df.drop(columns=['ID'], inplace=True)

# Checking for duplicates after dropping ID column
df.duplicated().sum()

np.int64(35)

- We have 35 duplicate rows in the dataset, which is very less compared to the total number of rows (30,000). So we will drop these duplicate rows.

In [11]:
# Removing duplicate rows
df = df.drop_duplicates()

# Checking the shape of the DataFrame after removing duplicates and dropping ID column
df.shape

(29965, 24)

- After cleaning the dataset now we have 29,965 rows and 24 columns.

# **2. Handling Outliers**

In [12]:
# Function to detect Outliers in Numerical Variables using IQR Method
def detect_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)][column]
    return len(outliers), lower_bound, upper_bound

# Creating a copy if df to avoid modifying the original DataFrame
df1 = df.copy()

for col in numerical_columns:
    outlier_count, lower, upper = detect_outliers(df1, col)
    df1 = df1[(df1[col] >= lower) & (df1[col] <= upper)]
df1.reset_index(drop=True, inplace=True)

# Printing the shape of df1 after removing outliers
df1.shape

(14455, 24)

- If we remove outliers, we will lose more than 50% of the data, so we will not remove outliers. Instead, we can apply log transformation for skewed numerical variables or use tree-based models which are robust to skewed distributions and outliers.

# **3 Binning Age Variable**
- We will bin the `Age` variable into age groups to reduce noise and improve model performance.

In [13]:
# Binning the Age variable into age groups
df['Age_Groups'] = pd.cut(df['Age'], bins=[20, 25, 30, 35, 40, 45, 50, 55, 60, np.inf],
                             labels=['20-25', '25-30', '30-35', '35-40', '40-45', '45-50', '50-55', '55-60', '60+'],
                             right=False)

# Dropping the Age column as it is no longer needed
df.drop(columns=['Age'], inplace=True)

# Displaying the counts of each age groups
df['Age_Groups'].value_counts().sort_index()

Age_Groups
20-25    2683
25-30    6920
30-35    6071
35-40    5155
40-45    3853
45-50    2603
50-55    1627
55-60     714
60+       339
Name: count, dtype: int64

- We grouped the Age variable into age groups to capture more meaningful patterns instead of using it as a continuous variable.
- **Insights**: Most of the credit card owners are in the age group of 25-30, followed by 30-35 and 35-40. There are very few credit card owners in the age groups above 60.

In [14]:
# Calculating ratios of default payment next month by age groups
age_ratios = df.groupby('Age_Groups')[target_column].value_counts(normalize=True).unstack().fillna(0) * 100
age_ratios = age_ratios.rename(columns={0: 'No Default', 1: 'Default'})
age_ratios['Total_Customers'] = df['Age_Groups'].value_counts()

# Displaying the ratios
age_ratios.round(2)

C:\Users\vinee\AppData\Local\Temp\ipykernel_23272\102376293.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_ratios = df.groupby('Age_Groups')[target_column].value_counts(normalize=True).unstack().fillna(0) * 100


default_payment_next_month  No Default  Default  Total_Customers
Age_Groups                                                      
20-25                            72.83    27.17             2683
25-30                            78.80    21.20             6920
30-35                            80.66    19.34             6071
35-40                            78.64    21.36             5155
40-45                            77.55    22.45             3853
45-50                            76.33    23.67             2603
50-55                            75.60    24.40             1627
55-60                            74.09    25.91              714
60+                              71.68    28.32              339

- As we can see from the above table, age above 60 has the highest chance of defaulting on payment next month (around 30%), followed by age group 20-25 (around 27%). Ages between 25-50 have a lower chance of defaulting, with the lowest chance in the age group 30-35 (around 20%).

In [15]:
# Creating new directory to save csv output files
csv_output_dir = os.path.join(data_dir, "output_csv")
os.makedirs(csv_output_dir, exist_ok=True)

# Saving the cleaned DataFrame to a new CSV file
cleaned_data_path = os.path.join(csv_output_dir, "cleaned_data.csv")
df.to_csv(cleaned_data_path, index=False)

# **4. Encoding Categorical Variables**
- We will encode ordinal categorical variables using OrdinalEncoder and nominal categorical variables using OneHotEncoder to prepare them for modeling.

In [16]:
# Loading the cleaned DataFrame and saving it to a new variable df1
df1 = pd.read_csv(cleaned_data_path)
df1.head()

Credit_Limit  Gender  Education_Level  Marital_Status  Sept_Pay  Aug_Pay  \
0       20000.0       2                2               1         2        2   
1      120000.0       2                2               2        -1        2   
2       90000.0       2                2               2         0        0   
3       50000.0       2                2               1         0        0   
4       50000.0       1                2               1        -1        0   

   July_Pay  June_Pay  May_Pay  Apr_Pay  ...  May_Bill_Amt  Apr_Bill_Amt  \
0        -1        -1       -2       -2  ...           0.0           0.0   
1         0         0        0        2  ...        3455.0        3261.0   
2         0         0        0        0  ...       14948.0       15549.0   
3         0         0        0        0  ...       28959.0       29547.0   
4        -1         0        0        0  ...       19146.0       19131.0   

   Sept_Pay_Amt  Aug_Pay_Amt  July_Pay_Amt  June_Pay_Amt  May_Pay_Amt  \
0           0.0        689.0           0.0           0.0          0.0   
1           0.0       1000.0        1000.0        1000.0          0.0   
2        1518.0       1500.0        1000.0        1000.0       1000.0   
3        2000.0       2019.0        1200.0        1100.0       1069.0   
4        2000.0      36681.0       10000.0        9000.0        689.0   

   Apr_Pay_Amt  default_payment_next_month  Age_Groups  
0          0.0                           1       20-25  
1       2000.0                           1       25-30  
2       5000.0                           0       30-35  
3       1000.0                           0       35-40  
4        679.0                           0       55-60  

[5 rows x 24 columns]

In [17]:
# Applying Ordinal Encoding on Age Groups column
from sklearn.preprocessing import OrdinalEncoder

# Initializing the OrdinalEncoder
ordinal_encoder = OrdinalEncoder()

# Encoding the 'Age_Groups' column
df1['Age_Groups'] = ordinal_encoder.fit_transform(df1[['Age_Groups']]).astype(int)

# Displaying the first few rows of the DataFrame after encoding
df1.head()

Credit_Limit  Gender  Education_Level  Marital_Status  Sept_Pay  Aug_Pay  \
0       20000.0       2                2               1         2        2   
1      120000.0       2                2               2        -1        2   
2       90000.0       2                2               2         0        0   
3       50000.0       2                2               1         0        0   
4       50000.0       1                2               1        -1        0   

   July_Pay  June_Pay  May_Pay  Apr_Pay  ...  May_Bill_Amt  Apr_Bill_Amt  \
0        -1        -1       -2       -2  ...           0.0           0.0   
1         0         0        0        2  ...        3455.0        3261.0   
2         0         0        0        0  ...       14948.0       15549.0   
3         0         0        0        0  ...       28959.0       29547.0   
4        -1         0        0        0  ...       19146.0       19131.0   

   Sept_Pay_Amt  Aug_Pay_Amt  July_Pay_Amt  June_Pay_Amt  May_Pay_Amt  \
0           0.0        689.0           0.0           0.0          0.0   
1           0.0       1000.0        1000.0        1000.0          0.0   
2        1518.0       1500.0        1000.0        1000.0       1000.0   
3        2000.0       2019.0        1200.0        1100.0       1069.0   
4        2000.0      36681.0       10000.0        9000.0        689.0   

   Apr_Pay_Amt  default_payment_next_month  Age_Groups  
0          0.0                           1           0  
1       2000.0                           1           1  
2       5000.0                           0           2  
3       1000.0                           0           3  
4        679.0                           0           7  

[5 rows x 24 columns]

- Mapped Age Groups categories to numerical value using OrdinalEncoder for modelling.

In [18]:
# Nominal categories for categorical variables
nominal_categories = ['Gender', 'Marital_Status']

# Transforming nominal categorical variables into numerical values using one-hot encoding
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(sparse_output=False, drop='first')

# Applying one-hot encoding to the nominal categorical variables
encoded_features = encoder.fit_transform(df1[nominal_categories]).astype(int)

# Creating a DataFrame from the encoded features
encoded_df = pd.DataFrame(encoded_features, columns=encoder.get_feature_names_out(nominal_categories))

# Concatenating the encoded DataFrame with the original DataFrame (excluding the original nominal columns)
df1 = pd.concat([df1.drop(columns=nominal_categories), encoded_df], axis=1)

# Displaying the first few rows of the updated DataFrame with one-hot encoded features
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  Aug_Pay_Amt  \
0       -2       -2         3913.0        3102.0  ...        689.0   
1        0        2         2682.0        1725.0  ...       1000.0   
2        0        0        29239.0       14027.0  ...       1500.0   
3        0        0        46990.0       48233.0  ...       2019.0   
4        0        0         8617.0        5670.0  ...      36681.0   

   July_Pay_Amt  June_Pay_Amt  May_Pay_Amt  Apr_Pay_Amt  \
0           0.0           0.0          0.0          0.0   
1        1000.0        1000.0          0.0       2000.0   
2        1000.0        1000.0       1000.0       5000.0   
3        1200.0        1100.0       1069.0       1000.0   
4       10000.0        9000.0        689.0        679.0   

   default_payment_next_month  Age_Groups  Gender_2  Marital_Status_2  \
0                           1           0         1                 0   
1                           1           1         1                 1   
2                           0           2         1                 1   
3                           0           3         1                 0   
4                           0           7         0                 0   

   Marital_Status_3  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 25 columns]

- We have transformed all the nominal categorical variables into numerical values using one-hot encoding converting them into binary columns. This will allow us to use these variables in our machine learning models effectively.
- We have also dropped 1 column from one-hot encoding to avoid the dummy variable trap, as it is not needed for modeling.

# **5. Feature Engineering**

**Creating New Features**
- We will create new features based on the existing features to improve the predictive power of the model.

In [19]:
# Calculating Average Bill Amount over 6 months
df1['Avg_Bill_Amt'] = df1[['Sept_Bill_Amt', 'Aug_Bill_Amt', 'July_Bill_Amt', 'June_Bill_Amt', 'May_Bill_Amt', 'Apr_Bill_Amt']].mean(axis=1).round(2)

# Displaying the first few rows to verify the new column
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  July_Pay_Amt  \
0       -2       -2         3913.0        3102.0  ...           0.0   
1        0        2         2682.0        1725.0  ...        1000.0   
2        0        0        29239.0       14027.0  ...        1000.0   
3        0        0        46990.0       48233.0  ...        1200.0   
4        0        0         8617.0        5670.0  ...       10000.0   

   June_Pay_Amt  May_Pay_Amt  Apr_Pay_Amt  default_payment_next_month  \
0           0.0          0.0          0.0                           1   
1        1000.0          0.0       2000.0                           1   
2        1000.0       1000.0       5000.0                           0   
3        1100.0       1069.0       1000.0                           0   
4        9000.0        689.0        679.0                           0   

   Age_Groups  Gender_2  Marital_Status_2  Marital_Status_3  Avg_Bill_Amt  
0           0         1                 0                 0       1284.00  
1           1         1                 1                 0       2846.17  
2           2         1                 1                 0      16942.17  
3           3         1                 0                 0      38555.67  
4           7         0                 0                 0      18223.17  

[5 rows x 26 columns]

- Created new feature `Avg_Bill_Amt` which is the average of all bill amounts over the six months. This feature captures the overall spending behavior of the customer.

In [20]:
# Calculating Average Payment amount over 6 months
df1['Avg_Pay_Amt'] = df1[['Sept_Pay_Amt', 'Aug_Pay_Amt', 'July_Pay_Amt', 'June_Pay_Amt', 'May_Pay_Amt', 'Apr_Pay_Amt']].mean(axis=1).round(2)

# Displaying the first few rows to verify the new column
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  June_Pay_Amt  \
0       -2       -2         3913.0        3102.0  ...           0.0   
1        0        2         2682.0        1725.0  ...        1000.0   
2        0        0        29239.0       14027.0  ...        1000.0   
3        0        0        46990.0       48233.0  ...        1100.0   
4        0        0         8617.0        5670.0  ...        9000.0   

   May_Pay_Amt  Apr_Pay_Amt  default_payment_next_month  Age_Groups  Gender_2  \
0          0.0          0.0                           1           0         1   
1          0.0       2000.0                           1           1         1   
2       1000.0       5000.0                           0           2         1   
3       1069.0       1000.0                           0           3         1   
4        689.0        679.0                           0           7         0   

   Marital_Status_2  Marital_Status_3  Avg_Bill_Amt  Avg_Pay_Amt  
0                 0                 0       1284.00       114.83  
1                 1                 0       2846.17       833.33  
2                 1                 0      16942.17      1836.33  
3                 0                 0      38555.67      1398.00  
4                 0                 0      18223.17      9841.50  

[5 rows x 27 columns]

- Created new feature `Avg_Pay_Amt` which is the average of all payment amounts over the six months. This feature captures the overall payment behavior of the customer.

In [21]:
# Calculating Average Payment Delay Score (average of payment status)
df1['Avg_Delay_Score'] = df1[['Sept_Pay', 'Aug_Pay', 'July_Pay', 'June_Pay', 'May_Pay', 'Apr_Pay']].mean(axis=1).round(2)

# Displaying the first few rows to verify the new column
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  May_Pay_Amt  \
0       -2       -2         3913.0        3102.0  ...          0.0   
1        0        2         2682.0        1725.0  ...          0.0   
2        0        0        29239.0       14027.0  ...       1000.0   
3        0        0        46990.0       48233.0  ...       1069.0   
4        0        0         8617.0        5670.0  ...        689.0   

   Apr_Pay_Amt  default_payment_next_month  Age_Groups  Gender_2  \
0          0.0                           1           0         1   
1       2000.0                           1           1         1   
2       5000.0                           0           2         1   
3       1000.0                           0           3         1   
4        679.0                           0           7         0   

   Marital_Status_2  Marital_Status_3  Avg_Bill_Amt  Avg_Pay_Amt  \
0                 0                 0       1284.00       114.83   
1                 1                 0       2846.17       833.33   
2                 1                 0      16942.17      1836.33   
3                 0                 0      38555.67      1398.00   
4                 0                 0      18223.17      9841.50   

   Avg_Delay_Score  
0            -0.33  
1             0.50  
2             0.00  
3             0.00  
4            -0.33  

[5 rows x 28 columns]

- Created new feature `Avg_Delay_Score` which is the average of all delay scores over the six months. This feature captures the overall delay behavior of the customer.

In [22]:
# Calculating Payment to Bill Ratios by Month
df1['Sept_Pay_Bill_Ratio'] = np.where(df1['Sept_Bill_Amt'] != 0, df1['Sept_Pay_Amt'] / df1['Sept_Bill_Amt'], 0).round(2)
df1['Aug_Pay_Bill_Ratio'] = np.where(df1['Aug_Bill_Amt'] != 0, df1['Aug_Pay_Amt'] / df1['Aug_Bill_Amt'], 0).round(2)
df1['July_Pay_Bill_Ratio'] = np.where(df1['July_Bill_Amt'] != 0, df1['July_Pay_Amt'] / df1['July_Bill_Amt'], 0).round(2)
df1['June_Pay_Bill_Ratio'] = np.where(df1['June_Bill_Amt'] != 0, df1['June_Pay_Amt'] / df1['June_Bill_Amt'], 0).round(2)
df1['May_Pay_Bill_Ratio'] = np.where(df1['May_Bill_Amt'] != 0, df1['May_Pay_Amt'] / df1['May_Bill_Amt'], 0).round(2)
df1['Apr_Pay_Bill_Ratio'] = np.where(df1['Apr_Bill_Amt'] != 0, df1['Apr_Pay_Amt'] / df1['Apr_Bill_Amt'], 0).round(2)

# Calculating Average Payment to Bill Ratio for the last 6 months
df1['Avg_Pay_Bill_Ratio'] = df1[['Sept_Pay_Bill_Ratio', 'Aug_Pay_Bill_Ratio', 'July_Pay_Bill_Ratio', 
                                       'June_Pay_Bill_Ratio', 'May_Pay_Bill_Ratio', 'Apr_Pay_Bill_Ratio']].mean(axis=1).round(2)


# Displaying the first few rows of the updated DataFrame
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  Avg_Bill_Amt  \
0       -2       -2         3913.0        3102.0  ...       1284.00   
1        0        2         2682.0        1725.0  ...       2846.17   
2        0        0        29239.0       14027.0  ...      16942.17   
3        0        0        46990.0       48233.0  ...      38555.67   
4        0        0         8617.0        5670.0  ...      18223.17   

   Avg_Pay_Amt  Avg_Delay_Score  Sept_Pay_Bill_Ratio  Aug_Pay_Bill_Ratio  \
0       114.83            -0.33                 0.00                0.22   
1       833.33             0.50                 0.00                0.58   
2      1836.33             0.00                 0.05                0.11   
3      1398.00             0.00                 0.04                0.04   
4      9841.50            -0.33                 0.23                6.47   

   July_Pay_Bill_Ratio  June_Pay_Bill_Ratio  May_Pay_Bill_Ratio  \
0                 0.00                 0.00                0.00   
1                 0.37                 0.31                0.00   
2                 0.07                 0.07                0.07   
3                 0.02                 0.04                0.04   
4                 0.28                 0.43                0.04   

   Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  
0                0.00                0.04  
1                0.61                0.31  
2                0.32                0.12  
3                0.03                0.04  
4                0.04                1.25  

[5 rows x 35 columns]

- Created `Sept_Pay_Bill_Ratio` to `Apr_Pay_Bill_Ratio` which is the ratio of payment amount to bill amount for each month. This feature captures the payment behavior of the customer in relation to their bill amounts.
- Created `Avg_Pay_Bill_Ratio` which is the average of all payment to bill ratios over the six months. This feature captures the overall payment behavior of the customer in relation to their bill amounts.

In [23]:
# Calculating Credit utilization ratio of each month
df1['Sept_Credit_Utilization_Ratio'] = (df1['Sept_Pay_Amt'] / df1['Credit_Limit']).round(2)
df1['Aug_Credit_Utilization_Ratio'] = (df1['Aug_Pay_Amt'] / df1['Credit_Limit']).round(2)
df1['July_Credit_Utilization_Ratio'] = (df1['July_Pay_Amt'] / df1['Credit_Limit']).round(2)
df1['June_Credit_Utilization_Ratio'] = (df1['June_Pay_Amt'] / df1['Credit_Limit']).round(2)
df1['May_Credit_Utilization_Ratio'] = (df1['May_Pay_Amt'] / df1['Credit_Limit']).round(2)
df1['Apr_Credit_Utilization_Ratio'] = (df1['Apr_Pay_Amt'] / df1['Credit_Limit']).round(2)

# Calculating Average credit utilization ratio for the last 6 months
df1['Avg_Credit_Utilization_Ratio'] = df1[['Sept_Credit_Utilization_Ratio', 'Aug_Credit_Utilization_Ratio', 
                                             'July_Credit_Utilization_Ratio', 'June_Credit_Utilization_Ratio', 
                                             'May_Credit_Utilization_Ratio', 'Apr_Credit_Utilization_Ratio']].mean(axis=1).round(2)

# Displaying the first few rows of the updated DataFrame
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  May_Pay_Bill_Ratio  \
0       -2       -2         3913.0        3102.0  ...                0.00   
1        0        2         2682.0        1725.0  ...                0.00   
2        0        0        29239.0       14027.0  ...                0.07   
3        0        0        46990.0       48233.0  ...                0.04   
4        0        0         8617.0        5670.0  ...                0.04   

   Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  Sept_Credit_Utilization_Ratio  \
0                0.00                0.04                           0.00   
1                0.61                0.31                           0.00   
2                0.32                0.12                           0.02   
3                0.03                0.04                           0.04   
4                0.04                1.25                           0.04   

   Aug_Credit_Utilization_Ratio  July_Credit_Utilization_Ratio  \
0                          0.03                           0.00   
1                          0.01                           0.01   
2                          0.02                           0.01   
3                          0.04                           0.02   
4                          0.73                           0.20   

   June_Credit_Utilization_Ratio  May_Credit_Utilization_Ratio  \
0                           0.00                          0.00   
1                           0.01                          0.00   
2                           0.01                          0.01   
3                           0.02                          0.02   
4                           0.18                          0.01   

   Apr_Credit_Utilization_Ratio  Avg_Credit_Utilization_Ratio  
0                          0.00                          0.00  
1                          0.02                          0.01  
2                          0.06                          0.02  
3                          0.02                          0.03  
4                          0.01                          0.19  

[5 rows x 42 columns]

- Created `Sept_Credit_Utilization_Ratio` to `Apr_Credit_Utilization_Ratio` which is the ratio of credit limit to bill amount for each month. This feature captures the credit utilization behavior of the customer.
- Created `Avg_Credit_Utilization_Ratio` which is the average of all credit utilization ratios over the six months. This feature captures the overall credit utilization behavior of the customer.

In [24]:
# Displaying the first few rows of the updated DataFrame with new features
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  May_Pay_Bill_Ratio  \
0       -2       -2         3913.0        3102.0  ...                0.00   
1        0        2         2682.0        1725.0  ...                0.00   
2        0        0        29239.0       14027.0  ...                0.07   
3        0        0        46990.0       48233.0  ...                0.04   
4        0        0         8617.0        5670.0  ...                0.04   

   Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  Sept_Credit_Utilization_Ratio  \
0                0.00                0.04                           0.00   
1                0.61                0.31                           0.00   
2                0.32                0.12                           0.02   
3                0.03                0.04                           0.04   
4                0.04                1.25                           0.04   

   Aug_Credit_Utilization_Ratio  July_Credit_Utilization_Ratio  \
0                          0.03                           0.00   
1                          0.01                           0.01   
2                          0.02                           0.01   
3                          0.04                           0.02   
4                          0.73                           0.20   

   June_Credit_Utilization_Ratio  May_Credit_Utilization_Ratio  \
0                           0.00                          0.00   
1                           0.01                          0.00   
2                           0.01                          0.01   
3                           0.02                          0.02   
4                           0.18                          0.01   

   Apr_Credit_Utilization_Ratio  Avg_Credit_Utilization_Ratio  
0                          0.00                          0.00  
1                          0.02                          0.01  
2                          0.06                          0.02  
3                          0.02                          0.03  
4                          0.01                          0.19  

[5 rows x 42 columns]

In [25]:
# Checking the shape of the DataFrame after adding new features
df1.shape

(29965, 42)

- Now we have 29,965 rows and 42 columns in the dataset after performing all the preprocessing steps.

In [26]:
# Moving default_payment_next_month column to the last position
cols = [col for col in df1.columns if col != 'default_payment_next_month']
cols.append('default_payment_next_month')
df1 = df1[cols]

# Display the DataFrame to confirm the change
df1.head()

Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
0       20000.0                2         2        2        -1        -1   
1      120000.0                2        -1        2         0         0   
2       90000.0                2         0        0         0         0   
3       50000.0                2         0        0         0         0   
4       50000.0                2        -1        0        -1         0   

   May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  Apr_Pay_Bill_Ratio  \
0       -2       -2         3913.0        3102.0  ...                0.00   
1        0        2         2682.0        1725.0  ...                0.61   
2        0        0        29239.0       14027.0  ...                0.32   
3        0        0        46990.0       48233.0  ...                0.03   
4        0        0         8617.0        5670.0  ...                0.04   

   Avg_Pay_Bill_Ratio  Sept_Credit_Utilization_Ratio  \
0                0.04                           0.00   
1                0.31                           0.00   
2                0.12                           0.02   
3                0.04                           0.04   
4                1.25                           0.04   

   Aug_Credit_Utilization_Ratio  July_Credit_Utilization_Ratio  \
0                          0.03                           0.00   
1                          0.01                           0.01   
2                          0.02                           0.01   
3                          0.04                           0.02   
4                          0.73                           0.20   

   June_Credit_Utilization_Ratio  May_Credit_Utilization_Ratio  \
0                           0.00                          0.00   
1                           0.01                          0.00   
2                           0.01                          0.01   
3                           0.02                          0.02   
4                           0.18                          0.01   

   Apr_Credit_Utilization_Ratio  Avg_Credit_Utilization_Ratio  \
0                          0.00                          0.00   
1                          0.02                          0.01   
2                          0.06                          0.02   
3                          0.02                          0.03   
4                          0.01                          0.19   

   default_payment_next_month  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0  

[5 rows x 42 columns]

In [27]:
# saving the updated DataFrame with new features to a new CSV file
final_data_path = os.path.join(csv_output_dir, "final_preprocessed_data.csv")
df1.to_csv(final_data_path, index=False)

# **6. Data Splitting**
- We will split the dataset into training and testing sets using an 75-25 split. This will allow us to train the model on a larger portion of the data and evaluate its performance on unseen data.

In [28]:
# Loading the final preprocessed DataFrame
df2 = pd.read_csv(final_data_path)

In [29]:
# Split data into X (features) and y (target)
target_column = 'default_payment_next_month'
X = df2.drop(columns=[target_column])
y = df2[target_column]

# Displaying the shape of the features and target
print(f"Shape of features (X): {X.shape}")
print(f"Shape of target (y): {y.shape}")

Shape of features (X): (29965, 41)
Shape of target (y): (29965,)


In [30]:
# Spliting the dataset into training and testing sets
from sklearn.model_selection import train_test_split

# Splitting the dataset into training and testing sets with a 75-25 split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, stratify=y)

# Displaying the shapes of the training and testing sets
print(f"Shape of training features (X_train): {X_train.shape}")
print(f"Shape of testing features (X_test): {X_test.shape}")

Shape of training features (X_train): (22473, 41)
Shape of testing features (X_test): (7492, 41)


In [31]:
# Checking proportions of default payment next month in training data
y_train.value_counts(normalize=True).round(2)

default_payment_next_month
0    0.78
1    0.22
Name: proportion, dtype: float64

In [32]:
# Checking proportions of default payment next month in testing data
y_test.value_counts(normalize=True).round(2)

default_payment_next_month
0    0.78
1    0.22
Name: proportion, dtype: float64

- We have same proportion of target variable in both training and testing sets, which is good for model evaluation.

# **7. Feature Scaling (Standardization)**
- We will use **fit_transform** method for **X_train**, which will scale the features to have a mean of 0 and a standard deviation of 1. 
- We will use **transform** method for **X_test**, which will scale the features using the same parameters (mean and standard deviation) as X_train. This is important to ensure that the test data is scaled in the same way as the training data **to prevents data leakage** and ensures that the model is evaluated on the same scale as it was trained on.
- Feature scaling is optional for tree-based models like Random Forest and XGBoost, but it is essential for algorithms that are sensitive to the scale of the features, such as logistic regression and k-nearest neighbors.

In [33]:
# Feature scaling using StandardScaler
from sklearn.preprocessing import StandardScaler

# Initializing the StandardScaler
scaler = StandardScaler()

# Fit the scaler on X_train and transform both X_train and X_test
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Convert back to DataFrame with original column names and indices
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns, index=X_train.index)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns, index=X_test.index)

# Display first few rows of scaled training data
print(X_train_scaled.shape)
X_train_scaled.head()

(22473, 41)


Credit_Limit  Education_Level  Sept_Pay   Aug_Pay  July_Pay  June_Pay  \
2447      -0.286942         0.308263  0.010580  0.108000  0.137993  0.192481   
16962     -0.826903         0.308263  1.791635  1.781855  1.820638  1.918243   
5001      -0.672629         1.684695  0.010580  0.108000  0.137993  0.192481   
5861       1.178667         0.308263 -0.879947 -0.728927 -1.544652 -1.533282   
16696     -0.364079         0.308263 -0.879947  0.108000  0.137993  0.192481   

        May_Pay   Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  \
2447   0.239828  0.259337      -0.452249     -0.461789  ...   
16962  0.239828  0.259337      -0.148335     -0.111258  ...   
5001   0.239828  0.259337      -0.091411     -0.078527  ...   
5861  -1.541873 -1.495054      -0.196536     -0.288899  ...   
16696  0.239828  0.259337       0.834479      0.917161  ...   

       May_Pay_Bill_Ratio  Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  \
2447             0.013111            0.008450            0.018362   
16962            0.011626            0.007773            0.017899   
5001             0.011626            0.007547            0.018131   
5861             0.021724            0.050901            0.027323   
16696            0.011032            0.006870            0.017744   

       Sept_Credit_Utilization_Ratio  Aug_Credit_Utilization_Ratio  \
2447                       -0.250428                     -0.257562   
16962                      -0.150985                     -0.108932   
5001                       -0.250428                      0.559906   
5861                        0.445672                     -0.034616   
16696                      -0.150985                     -0.034616   

       July_Credit_Utilization_Ratio  June_Credit_Utilization_Ratio  \
2447                       -0.301519                      -0.158567   
16962                      -0.407383                      -0.059798   
5001                       -0.195655                      -0.158567   
5861                       -0.301519                      -0.059798   
16696                      -0.195655                      -0.257335   

       May_Credit_Utilization_Ratio  Apr_Credit_Utilization_Ratio  \
2447                      -0.164245                     -0.259512   
16962                     -0.052211                     -0.058064   
5001                      -0.164245                     -0.158788   
5861                      -0.164245                      0.042660   
16696                     -0.388313                     -0.360236   

       Avg_Credit_Utilization_Ratio  
2447                      -0.348635  
16962                     -0.348635  
5001                       0.018175  
5861                       0.018175  
16696                     -0.348635  

[5 rows x 41 columns]

In [34]:
# Display first few rows of scaled testing data
print(X_test_scaled.shape)
X_test_scaled.head()

(7492, 41)


Credit_Limit  Education_Level  Sept_Pay   Aug_Pay  July_Pay  June_Pay  \
7356      -0.132667        -1.068170  0.010580  0.108000  0.137993  0.192481   
12259      0.947255         1.684695 -0.879947 -0.728927 -0.703330 -0.670400   
29345      0.947255        -1.068170  0.010580  0.108000  0.137993  1.918243   
24586      1.487217        -1.068170 -0.879947 -0.728927  1.820638 -0.670400   
14969      2.412865         0.308263 -0.879947 -0.728927 -0.703330 -0.670400   

        May_Pay   Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  \
7356   0.239828  0.259337       1.358983      1.434606  ...   
12259 -0.651023 -0.617858      -0.695461     -0.689077  ...   
29345  2.021529  2.013728      -0.613313     -0.591334  ...   
24586 -0.651023  0.259337      -0.697268     -0.678603  ...   
14969 -0.651023 -0.617858      -0.676796     -0.669864  ...   

       May_Pay_Bill_Ratio  Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  \
7356             0.012220            0.006870            0.017899   
12259            0.025882            0.029450            0.025314   
29345            0.011032            0.008676            0.018440   
24586            0.011032            0.029450            0.035433   
14969            0.025882            0.029450            0.025314   

       Sept_Credit_Utilization_Ratio  Aug_Credit_Utilization_Ratio  \
7356                       -0.051542                     -0.108932   
12259                      -0.449313                     -0.331877   
29345                      -0.449313                     -0.257562   
24586                      -0.449313                     -0.331877   
14969                      -0.449313                     -0.331877   

       July_Credit_Utilization_Ratio  June_Credit_Utilization_Ratio  \
7356                       -0.089791                       0.038970   
12259                      -0.407383                      -0.356104   
29345                      -0.407383                      -0.356104   
24586                      -0.301519                      -0.356104   
14969                      -0.407383                      -0.356104   

       May_Credit_Utilization_Ratio  Apr_Credit_Utilization_Ratio  \
7356                       0.395926                     -0.360236   
12259                     -0.388313                     -0.360236   
29345                     -0.388313                     -0.360236   
24586                     -0.388313                     -0.360236   
14969                     -0.388313                     -0.360236   

       Avg_Credit_Utilization_Ratio  
7356                       0.018175  
12259                     -0.715444  
29345                     -0.715444  
24586                     -0.715444  
14969                     -0.715444  

[5 rows x 41 columns]

- We have performed feature scaling on the training and testing sets using StandardScaler. 
- We have ensured that there are no data leaks by using the fit_transform method on the training set and transform method on the testing set. This ensures that the scaling parameters (mean and standard deviation) are derived only from the training data and applied to the test data.

# **8. Saving Train and Test Data**

In [35]:
# Concatenating the training sets into a single DataFrame for saving
training_data = pd.concat([X_train_scaled, y_train], axis=1)
print(f"Shape of training data: {training_data.shape}")
training_data.head()

Shape of training data: (22473, 42)


Credit_Limit  Education_Level  Sept_Pay   Aug_Pay  July_Pay  June_Pay  \
2447      -0.286942         0.308263  0.010580  0.108000  0.137993  0.192481   
16962     -0.826903         0.308263  1.791635  1.781855  1.820638  1.918243   
5001      -0.672629         1.684695  0.010580  0.108000  0.137993  0.192481   
5861       1.178667         0.308263 -0.879947 -0.728927 -1.544652 -1.533282   
16696     -0.364079         0.308263 -0.879947  0.108000  0.137993  0.192481   

        May_Pay   Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  \
2447   0.239828  0.259337      -0.452249     -0.461789  ...   
16962  0.239828  0.259337      -0.148335     -0.111258  ...   
5001   0.239828  0.259337      -0.091411     -0.078527  ...   
5861  -1.541873 -1.495054      -0.196536     -0.288899  ...   
16696  0.239828  0.259337       0.834479      0.917161  ...   

       Apr_Pay_Bill_Ratio  Avg_Pay_Bill_Ratio  Sept_Credit_Utilization_Ratio  \
2447             0.008450            0.018362                      -0.250428   
16962            0.007773            0.017899                      -0.150985   
5001             0.007547            0.018131                      -0.250428   
5861             0.050901            0.027323                       0.445672   
16696            0.006870            0.017744                      -0.150985   

       Aug_Credit_Utilization_Ratio  July_Credit_Utilization_Ratio  \
2447                      -0.257562                      -0.301519   
16962                     -0.108932                      -0.407383   
5001                       0.559906                      -0.195655   
5861                      -0.034616                      -0.301519   
16696                     -0.034616                      -0.195655   

       June_Credit_Utilization_Ratio  May_Credit_Utilization_Ratio  \
2447                       -0.158567                     -0.164245   
16962                      -0.059798                     -0.052211   
5001                       -0.158567                     -0.164245   
5861                       -0.059798                     -0.164245   
16696                      -0.257335                     -0.388313   

       Apr_Credit_Utilization_Ratio  Avg_Credit_Utilization_Ratio  \
2447                      -0.259512                     -0.348635   
16962                     -0.058064                     -0.348635   
5001                      -0.158788                      0.018175   
5861                       0.042660                      0.018175   
16696                     -0.360236                     -0.348635   

       default_payment_next_month  
2447                            0  
16962                           1  
5001                            0  
5861                            0  
16696                           0  

[5 rows x 42 columns]

In [36]:
# Concatenating the testing sets into a single DataFrame for saving
testing_data = pd.concat([X_test, y_test], axis=1)
print(f"Shape of testing data: {testing_data.shape}")
testing_data.head()

Shape of testing data: (7492, 42)


Credit_Limit  Education_Level  Sept_Pay  Aug_Pay  July_Pay  June_Pay  \
7356       150000.0                1         0        0         0         0   
12259      290000.0                3        -1       -1        -1        -1   
29345      290000.0                1         0        0         0         2   
24586      360000.0                1        -1       -1         2        -1   
14969      480000.0                2        -1       -1        -1        -1   

       May_Pay  Apr_Pay  Sept_Bill_Amt  Aug_Bill_Amt  ...  Apr_Pay_Bill_Ratio  \
7356         0        0       150426.0      151052.0  ...                0.00   
12259       -1       -1          396.0         396.0  ...                1.00   
29345        2        2         6395.0        7330.0  ...                0.08   
24586       -1        0          264.0        1139.0  ...                1.00   
14969       -1       -1         1759.0        1759.0  ...                1.00   

       Avg_Pay_Bill_Ratio  Sept_Credit_Utilization_Ratio  \
7356                 0.04                           0.04   
12259                1.00                           0.00   
29345                0.11                           0.00   
24586                2.31                           0.00   
14969                1.00                           0.00   

       Aug_Credit_Utilization_Ratio  July_Credit_Utilization_Ratio  \
7356                           0.03                           0.03   
12259                          0.00                           0.00   
29345                          0.01                           0.00   
24586                          0.00                           0.01   
14969                          0.00                           0.00   

       June_Credit_Utilization_Ratio  May_Credit_Utilization_Ratio  \
7356                            0.04                          0.07   
12259                           0.00                          0.00   
29345                           0.00                          0.00   
24586                           0.00                          0.00   
14969                           0.00                          0.00   

       Apr_Credit_Utilization_Ratio  Avg_Credit_Utilization_Ratio  \
7356                            0.0                          0.04   
12259                           0.0                          0.00   
29345                           0.0                          0.00   
24586                           0.0                          0.00   
14969                           0.0                          0.00   

       default_payment_next_month  
7356                            0  
12259                           0  
29345                           1  
24586                           0  
14969                           1  

[5 rows x 42 columns]

In [37]:
# Saving the training data to a new CSV file
train_data_path = os.path.join(csv_output_dir, "train_data.csv")
training_data.to_csv(train_data_path, index=False)

# Saving the testing data to a new CSV file
test_data_path = os.path.join(csv_output_dir, "test_data.csv")
testing_data.to_csv(test_data_path, index=False)

# **Summary of Data Preprocessing**

- **Column Renaming:** 
    - Renamed columns for clarity and consistency.

- **Missing Value Handling:** 
    - Replaced missing values '0' in `Marital_Status` with the mode '2'(Single).
    - Replaced 4, 5, and 6 in Education_Level with 0 and consider them as others. This is because these values are not standard education levels and can be grouped together.
    - **Duplicate Removal:** 
        - Dropped 35 duplicate rows and unnecessary columns (like `ID`).

- **Outlier Handling:** 
    - Detected outliers but chose not to remove them to avoid excessive data loss; considered robust models instead.

- **Binning Age Variable:** 
    - Binned `Age` into groups to reduce noise and improve model performance.
    - **Insights:** Most credit card owners are in the age group of 25-30, with the highest default risk in ages above 60 and 20-25.

- **Encoding Categorical Variables:** 
    - Used OrdinalEncoder for ordinal variables (`Age_Groups`) and OneHotEncoder for nominal variables (`Gender`, `Marital_Status`, `Education_Level`).
    - Ensured no dummy variable trap by dropping one column from one-hot encoding.

- **Feature Engineering:** 
    - Created new feature `Avg_Bill_Amt` which is the average of all bill amounts over the six months. This feature captures the overall spending behavior of the customer.
    - Created new feature `Avg_Pay_Amt` which is the average of all payment amounts over the six months. This feature captures the overall payment behavior of the customer.
    - Created new feature `Avg_Delay_Score` which is the average of all delay scores over the six months. This feature captures the overall delay behavior of the customer.
    - Created `Sept_Pay_Bill_Ratio` to `Apr_Pay_Bill_Ratio` which is the ratio of payment amount to bill amount for each month. This feature captures the payment behavior of the customer in relation to their bill amounts.
    - Created `Avg_Pay_Bill_Ratio` which is the average of all payment to bill ratios over the six months. This feature captures the overall payment behavior of the customer in relation to their bill amounts.
    - Created `Sept_Credit_Utilization_Ratio` to `Apr_Credit_Utilization_Ratio` which is the ratio of credit limit to bill amount for each month. This feature captures the credit utilization behavior of the customer.
    - Created `Avg_Credit_Utilization_Ratio` which is the average of all credit utilization ratios over the six months. This feature captures the overall credit utilization behavior of the customer.

- **Data Splitting:** 
    - Split the dataset into training and testing sets using a 75-25 split, ensuring the same proportion of target variable in both sets for model evaluation.

- **Feature Scaling:** 
    - Standardized features using `StandardScaler` to ensure consistent scaling for modeling.
    - Ensured no data leakage by using `fit_transform` on training data and `transform` on testing data.

- **Saving Processed Data:** 
    - Saved the cleaned, feature-engineered, and scaled datasets for model training (`train_data.csv`) and evaluation (`test_data.csv`).